crates/shared/src/code_fetching.rs

```diff
@@ -0,0 +1,106 @@
+//! Module containing traits for abstracting Web3 operations so components can
+//! more easily be tested with mocked versions of these behaviours.
+
+use crate::Web3;
+use anyhow::Result;
+use cached::{Cached, SizedCache};
+use std::sync::{Arc, Mutex};
+use web3::types::{Bytes, H160};
+
+#[mockall::automock]
+#[async_trait::async_trait]
+pub trait CodeFetching: Send + Sync + 'static {
+    /// Fetches the code size at the specified address.
+    async fn code(&self, address: H160) -> Result<Bytes>;
+
+    /// Fetches the code for the specified address.
+    async fn code_size(&self, address: H160) -> Result<usize>;
+}
+
+#[async_trait::async_trait]
+impl CodeFetching for Web3 {
+    async fn code(&self, address: H160) -> Result<Bytes> {
+        Ok(self.eth().code(address, None).await?)
+    }
+
+    async fn code_size(&self, address: H160) -> Result<usize> {
+        Ok(self.code(address).await?.0.len())
+    }
+}
+
+pub struct CachedCodeFetcher {
+    inner: Arc<dyn CodeFetching>,
+    cache: Mutex<SizedCache<H160, Bytes>>,
+}
+
+impl CachedCodeFetcher {
+    const CACHE_SIZE: usize = 1_000;
+
+    pub fn new(inner: Arc<dyn CodeFetching>) -> Self {
+        Self {
+            inner,
+            cache: Mutex::new(SizedCache::with_size(Self::CACHE_SIZE)),
+        }
+    }
+
+    async fn cached_code<T, F>(&self, address: H160, handle: F) -> Result<T>
+    where
+        F: FnOnce(&Bytes) -> T,
```

Comment:
Yes, its because I didn't want to copy out the `Bytes` for the `code_size` call (we would clone a vector, to read its size, just to drop if afterwards).